lib/fuser.h

```diff
@@ -51,6 +52,87 @@ struct GateFused {
   std::vector<const Gate*> gates;
 };
 
+/**
+ * A base class for fuser classes with some common functions.
+ */
+template <typename IO, typename Gate>
+class Fuser {
+ protected:
+  using RGate = typename std::remove_pointer<Gate>::type;
+
+  static const RGate& GateToConstRef(const RGate& gate) {
+    return gate;
+  }
+
+  static const RGate& GateToConstRef(const RGate* gate) {
+    return *gate;
+  }
+
+  static std::vector<unsigned> MergeWithMeasurementTimes(
+      typename std::vector<Gate>::const_iterator gfirst,
+      typename std::vector<Gate>::const_iterator glast,
+      const std::vector<unsigned>& times) {
+    std::vector<unsigned> epochs;
+    epochs.reserve(glast - gfirst + times.size());
+
+    std::size_t last = 0;
+    unsigned max_time = 0;
+
+    for (auto gate_it = gfirst; gate_it < glast; ++gate_it) {
+      const auto& gate = GateToConstRef(*gate_it);
+
+      if (gate.time > max_time) {
+        max_time = gate.time;
+      }
+
+      if (epochs.size() > 0 && gate.time < epochs.back()) {
+        IO::errorf("gate crosses the time boundary.\n");
+        epochs.resize(0);
+        return epochs;
+      }
+
+      if (gate.kind == gate::kMeasurement) {
+        if (epochs.size() == 0 || epochs.back() < gate.time) {
+          if (!AddBoundary(gate.time, max_time, epochs)) {
+            epochs.resize(0);
+            return epochs;
+          }
+        }
+      }
+
+      if (last < times.size() && gate.time >= times[last]) {
```

Comment:
Is this `if` block necessary? Seems like the `while` condition is identical.

Reply:
You are right, thanks. Fixed.